### DISCUSSION TITLE
Standardised way to install plugins

### DISCUSSION THREAD

**Discussion by tombh** [nushell/nupm#106] [Q&A]
I'm new to Nushell so some of these questions might already be obvious or not relevant to this project.

Basically I'm wondering about the standardised way to install Nushell plugins. Specifically I'm wanting to install https://github.com/tesujimath/bash-env-nushell, which in turn depends on https://github.com/tesujimath/bash-env-json.

Ideally what I'd like is some unified config that I can place in my `config.nu` that both installs the plugin manager (if it's not already installed) and defines all the plugins that I use. Basically I don't want to have to remember all the commands I used to get setup the next time I install Nushell on a different system.

So, I think my questions are:
1. Can this project be self installed with zero manual user interaction?
2. Can this project install `bash-env-nushell` and `bash-env-json`? (I can't seem to find them in this project's repositories)
3. Even though `bash-env-nushell` can't be installed with `cargo install ...` is it still formally considered a Nushell plugin?

**Comment by fdncred** [comment]
The bash-env-nushell repo is pretty confusing. You look at the repo and it clearly says that `bash-env-nushell` is a nushell module, which is not a nushell plugin. However, if you look at the rust code in their repo, it looks like it has a nushell plugin named nu_plugin_bash_env, which automatically downloads bash-env-json when it's built and installed with `cargo install --path .` and maybe from crates.io https://crates.io/crates/nu_plugin_bash_env. I think the questions about the bash-env-nushell repo are best left to the author. My guess is that the answers to 1, 2, and 3 are yes but I'm not the author and don't use it.

**Comment by tombh** [comment]
Thank you.

Okay, I asked at the plugin repo: https://github.com/tesujimath/bash-env-nushell/issues/41

And so do you know how I might install `nupm` automatically from `config.nu` if it's on a new machine and detects that `nupm` isn't installed?

**Reply by fdncred** [reply]
I don't use nupm but it has installation instructions in the readme. https://github.com/nushell/nupm?tab=readme-ov-file#recycle-installation-toc

I don't think you'd want to install anything from within your config.nu file because it would run every time you start nushell.

**Reply by tombh** [reply]
Sure, thank you. Though I don't have a problem installing it. It's that I don't want to come back to the repo some months or years in the future when I've inevitably forgotten how I installed it. BASH and ZSH plugin managers usually have a nice conditional command that first checks if the plugin manager code is installed at a specific location and only do the installation on the first ever load of the config.
